Crash transitioning color with delay

Discovered during https://github.com/mapbox/mapbox-gl-native/pull/7711. 

If you have a sufficiently dark color, say a fill like `#333`, and apply runtime styling changes to black with any significant delay, an assertion will fail because the RBGA values drop below zero. 

Here is a core test proving the bug: 

https://github.com/mapbox/mapbox-gl-native/compare/core-bug-transition-color

If you change the commented-out `delay` argument from `0` to `1000` milliseconds ([this line](https://github.com/mapbox/mapbox-gl-native/compare/mapbox:255f09a...mapbox:b60b38d#diff-aa4701cd3cb986226f311d0d7d0c9df5R551)), the bug is triggered in the course of running `make run-test-Map.TransitionColor`. 

```
Note: Google Test filter = Map.TransitionColor
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Map
[ RUN      ] Map.TransitionColor
Assertion failed: (r_ >= 0.0f), function Color, file include/mbgl/util/color.hpp, line 16.
^Cmake: *** [run-test-Map.TransitionColor] Error 129
```

/cc @1ec5 @jfirebaugh @kkaefer

--- a/cmake/test-files.cmake
+++ b/cmake/test-files.cmake
@@ -55,6 +55,7 @@ set(MBGL_TEST_FILES
     test/src/mbgl/test/stub_file_source.hpp
     test/src/mbgl/test/stub_layer_observer.hpp
     test/src/mbgl/test/stub_style_observer.hpp
+    test/src/mbgl/test/stub_tile_observer.hpp
     test/src/mbgl/test/test.cpp
     test/src/mbgl/test/util.cpp
     test/src/mbgl/test/util.hpp
@@ -80,6 +81,7 @@ set(MBGL_TEST_FILES
     test/style/filter.test.cpp
     test/style/functions.test.cpp
     test/style/group_by_layout.test.cpp
+    test/style/paint_property.test.cpp
     test/style/source.test.cpp
     test/style/style.test.cpp
     test/style/style_layer.test.cpp

--- a/src/mbgl/style/paint_property.hpp
+++ b/src/mbgl/style/paint_property.hpp
@@ -45,6 +45,9 @@ public:
             // Transition from prior value is now complete.
             prior = {};
             return finalValue;
+        } else if (parameters.now < begin) {
+            // Transition hasn't started yet.
+            return prior->get().evaluate(parameters, defaultValue);
         } else {
             // Interpolate between recursively-calculated prior value and final.
             float t = std::chrono::duration<float>(parameters.now - begin) / (end - begin);

--- a/test/style/paint_property.test.cpp
+++ b/test/style/paint_property.test.cpp
@@ -1,0 +1,94 @@
+#include <mbgl/test/util.hpp>
+
+#include <mbgl/style/paint_property.hpp>
+
+using namespace mbgl;
+using namespace mbgl::style;
+using namespace std::literals::chrono_literals;
+
+TEST(UnevaluatedPaintProperty, EvaluateDefaultValue) {
+    UnevaluatedPaintProperty<float, PropertyEvaluator<float>> property;
+    ASSERT_EQ(0.0f, property.evaluate(PropertyEvaluationParameters(0), 0.0f));
+}
+
+TEST(UnevaluatedPaintProperty, EvaluateUntransitionedConstant) {
+    UnevaluatedPaintProperty<float, PropertyEvaluator<float>> property {
+        PropertyValue<float>(1.0f),
+        UnevaluatedPaintProperty<float, PropertyEvaluator<float>>(),
+        TransitionOptions(),
+        TimePoint::min()
+    };
+
+    ASSERT_EQ(1.0f, property.evaluate(PropertyEvaluationParameters(0), 0.0f));
+}
+
+TEST(UnevaluatedPaintProperty, EvaluateTransitionedConstantWithoutDelay) {
+    TransitionOptions transition;
+    transition.duration = { 1000ms };
+
+    UnevaluatedPaintProperty<float, PropertyEvaluator<float>> t0 {
+        PropertyValue<float>(0.0f),
+        UnevaluatedPaintProperty<float, PropertyEvaluator<float>>(),
+        TransitionOptions(),
+        TimePoint::min()
+    };
+
+    UnevaluatedPaintProperty<float, PropertyEvaluator<float>> t1 {
+        PropertyValue<float>(1.0f),
+        t0,
+        transition,
+        TimePoint::min()
+    };
+
+    auto evaluate = [&] (Duration delta) {
+        PropertyEvaluationParameters parameters {
+            0,
+            TimePoint::min() + delta,
+            ZoomHistory(),
+            Duration::zero()
+        };
+
+        return t1.evaluate(parameters, 0.0f);
+    };
+
+    ASSERT_FLOAT_EQ(0.0f, evaluate(0ms));
+    ASSERT_FLOAT_EQ(0.823099f, evaluate(500ms));
+    ASSERT_FLOAT_EQ(1.0f, evaluate(1500ms));
+}
+
+TEST(UnevaluatedPaintProperty, EvaluateTransitionedConstantWithDelay) {
+    TransitionOptions transition;
+    transition.delay = { 1000ms };
+    transition.duration = { 1000ms };
+
+    UnevaluatedPaintProperty<float, PropertyEvaluator<float>> t0 {
+        PropertyValue<float>(0.0f),
+        UnevaluatedPaintProperty<float, PropertyEvaluator<float>>(),
+        TransitionOptions(),
+        TimePoint::min()
+    };
+
+    UnevaluatedPaintProperty<float, PropertyEvaluator<float>> t1 {
+        PropertyValue<float>(1.0f),
+        t0,
+        transition,
+        TimePoint::min()
+    };
+
+    auto evaluate = [&] (Duration delta) {
+        PropertyEvaluationParameters parameters {
+            0,
+            TimePoint::min() + delta,
+            ZoomHistory(),
+            Duration::zero()
+        };
+
+        return t1.evaluate(parameters, 0.0f);
+    };
+
+    ASSERT_FLOAT_EQ(0.0f, evaluate(0ms));
+    ASSERT_FLOAT_EQ(0.0f, evaluate(500ms));
+    ASSERT_FLOAT_EQ(0.0f, evaluate(612ms));
+    ASSERT_FLOAT_EQ(0.823099f, evaluate(1500ms));
+    ASSERT_FLOAT_EQ(1.0f, evaluate(2500ms));
+}

